[Yi Hern] iP

YiHern-Lee opened the pull request:
# Stevie
> “Productivity is never an accident. It is always the result of a commitment to excellence, intelligent planning and focused effort.” – David Allen [(source)](https://www.goodreads.com/quotes/1133520-productivity-is-never-an-accident-it-is-always-the-result#:~:text=It%20is%20always%20the%20result,Meyer)

Stevie helps you stay productive by providing an engaging and simple way of recording and keeping track of upcoming tasks. Stevie is
* text-based
* easy to learn
* **SUPER** fast to use

## Getting Started
1. Download it from [here](https://github.com/YiHern-Lee/ip/releases/tag/A-Jar)
2. Save the `.jar` file locally on your computer
3. Open a terminal and run
```
java -r Stevie.jar
```
4. Start using Stevie to keep track of your upcoming tasks! 👍 

## Features
- [x] Adding and deleting upcoming *to-dos*, *deadlines*, and *events*
- [x] Saves and loads up tasks added with Stevie 
- [x] Marking tasks as done or not done
- [x] Searching tasks by name
- [ ] Graphical user interface (GUI) *(Coming Soon)*

### To change the tasks storage location 
Replace the class field `private static String path` in Stevie.java

```java
public class Stevie {
    private static String path = "src" + File.separator + "main"
                + File.separator + "data" + File.separator + "tasks.txt";
}
```

Commit:
Add Gradle support

LYHern committed:
Implemented basic dialogue feature for Duke. Duke echoes user's input.

LYHern committed:
Implemented Add Activities and List Activities features.

LYHern committed:
Add the ability to mark activities as done or undone. Users can now request to view helper text on how to operate Duke.

LYHern committed:
Add support for tracking three types of tasks: To-do, Event, Deadline. Renamed Activity to Task. Changed Task to an abstract class.

LYHern committed:
Use the input/output redirection technique to semi-automate the testing of Duke.

LYHern committed:
Duke has ability to handle errors such as incorrect inputs entered by the user.

LYHern committed:
Add support for deleting tasks from the list.

LYHern committed:
Added support for enums.

LYHern committed:
Rename Duke to Stevie, fix code style and add javadocs

LYHern committed:
Add saving and loading tasks feature

LYHern committed:
Add support for date-time formatting of events and deadlines

LYHern committed:
Merge branch 'branch-Level-7'

LYHern committed:
Arrange code in OOP manner

LYHern committed:
Refactor classes into packages

LYHern committed:
Add JUnit tests for some public methods

LYHern committed:
Add JUnit tests and refactor directories

LYHern committed:
Update .gitignore

LYHern committed:
Implement find feature to find tasks that contain certain keyword

LYHern committed:
Add Javadocs for classes and all public methods

LYHern committed:
Reformatted code to fit coding standards

LYHern committed:
Merge branch 'branch-A-JavaDoc'

LYHern committed:
Merge branch 'branch-A-CodingStandard', 'branch-A-JavaDoc", "branch-Level-9"

jamesyeap commented on src/main/java/stevie/StevieParser.java:
```diff
@@ -0,0 +1,100 @@
+package stevie;
+
+import stevie.command.*;
```
Maybe can change this to explicit imports? Like,
 ```java
// current
import stevie.command.*;

// changed to explicit import
import stevie.command.Command;
```

Changing this would satisfy the rule "Imported classes should always be listed explicitly." in the [Java coding standard](https://se-education.org/guides/conventions/java/intermediate.html#naming)

YiHern-Lee replied:
Thanks for pointing that out to me. I will take note of the import statements from now on.

jamesyeap commented on src/test/java/stevie/task/TaskCreatorTest.java:
```diff
@@ -0,0 +1,29 @@
+            assertEquals(new EventTask("cs2103 lecture", date).toString(),
+                    TaskCreator.create(TaskType.Event, false, "cs2103 lecture", date).toString());
+        } catch (TaskException | ParseException ex) {
+
```
Maybe can leave a comment here stating that nothing is to be done in the catch-block to be more explicit?

YiHern-Lee replied:
Good point! I will be sure to add a comment to identify portions of code that are not completed.

jamesyeap reviewed (commented):
Only managed to spot two deviations from the coding standard. Your code is pretty clean man 😁

howardwhw2 commented on src/main/java/stevie/StevieUi.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    private static String help() {
+        return "\"list\": to display your activities.\n" +
+                "\"bye\": to end our session.\n" +
+                "\"mark <i>\" to mark the i-th task as done.\n" +
+                "\"unmark <i>\" to unmark the i-th task as done.\n" +
+                "\"delete <i>\" to delete the i-th task.\n" +
+                "\"todo <task_name>\" to add a todo task.\n" +
+                "\"deadline <task_name> /by <date>\" to add a deadline.\n" +
+                "\"event <event_name> /at <date>\" to add an event.\n" +
+                "Date should in format of dd/mm/yyyy HH:mm";
```
Maybe the '+' operator should be placed after the line break?
```
     return "\"list\": to display your activities.\n" 
                + "\"bye\": to end our session.\n" 
                + "\"mark <i>\" to mark the i-th task as done.\n" 
                + "\"unmark <i>\" to unmark the i-th task as done.\n" 
                + "\"delete <i>\" to delete the i-th task.\n" 
                + "\"todo <task_name>\" to add a todo task.\n" 
                + "\"deadline <task_name> /by <date>\" to add a deadline.\n"
                + "\"event <event_name> /at <date>\" to add an event.\n" 
                + "Date should in format of dd/mm/yyyy HH:mm";
```
The "Place line break to improve readability" of the [coding standard](https://se-education.org/guides/conventions/java/intermediate.html) recommends line breaks to be done that way.

YiHern-Lee replied:
Thanks for pointing that out! I did not notice that when reviewing my own code.

howardwhw2 commented on src/main/java/stevie/task/TaskDataHandler.java:
```diff
@@ -0,0 +1,72 @@
+                taskList.add(TaskCreator.create(type, done, splits[2], date));
+            }
+            return taskList;
+        } catch (IOException | TaskException | ParseException ex) {
+            return new ArrayList<>();
+        }
```
Perhaps you might want to put in the JavaDoc that these exceptions are expected behaviour and why?

howardwhw2 commented on src/main/java/stevie/StevieUi.java:
```diff
@@ -0,0 +1,83 @@
+    public void greet() {
+        // Logo generated from:
+        // https://patorjk.com/software/taag/#p=display&f=Graffiti&t=Type%20Something%20
+        String logo = "  .--.--.       ___\n" +
+                " /  /    '.   ,--.'|_                        ,--,\n" +
+                "|  :  /`. /   |  | :,'                     ,--.'|\n" +
+                ";  |  |--`    :  : ' :                .---.|  |,\n" +
+                "|  :  ;_    .;__,'  /     ,---.     /.  ./|`--'_       ,---.\n" +
+                " \\  \\    `. |  |   |     /     \\  .-' . ' |,' ,'|     /     \\\n" +
+                "  `----.   \\:__,'| :    /    /  |/___/ \\: |'  | |    /    /  |\n" +
+                "  __ \\  \\  |  '  : |__ .    ' / |.   \\  ' .|  | :   .    ' / |\n" +
+                " /  /`--'  /  |  | '.'|'   ;   /| \\   \\   ''  : |__ '   ;   /|\n" +
+                "'--'.     /   ;  :    ;'   |  / |  \\   \\   |  | '.'|'   |  / |\n" +
+                "  `--'---'    |  ,   / |   :    |   \\   \\ |;  :    ;|   :    |\n" +
+                "               ---`-'   \\   \\  /     '---\" |  ,   /  \\   \\  /\n" +
+                "                         `----'             ---`-'    `----'   ";
```
similar suggestion with the '+' operator and great idea to get a logo 👍

howardwhw2 reviewed (commented):
Very neat code 💯

LYHern committed:
Fix coding standard violations

LYHern committed:
Merge branch 'add-gradle-support'

LYHern committed:
Add Gradle support for project

LYHern committed:
Add Gradle checkstyle support

LYHern committed:
Add support for GUI

LYHern committed:
Parser class: extract parse method into various classes with parse
method

Current parse method utilises multiple if-else clauses to match user
input to command.

Common attributes in some commands result in code duplication.

Extracting out the common behaviour into different parser classes allow
us to use polymorphism when parsing user input to commands. For example
adding of tasks can be abstracted into a TaskParser.

It will be easier to add parsers for new type of commands in the future.

LYHern committed:
Add assertions

LYHern committed:
Add support for github CI

YiHern-Lee committed:
Merge pull request #1 from YiHern-Lee/branch-A-CodeQuality

Extract parse method into various classes with parse method

LYHern committed:
Merge branch 'master' into branch-A-CI

LYHern committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/stevie/StevieParser.java
#	src/main/java/stevie/task/TaskDataHandler.java

YiHern-Lee committed:
Merge pull request #3 from YiHern-Lee/branch-A-Assertions

Add assertions

LYHern committed:
Merge branch 'master' into branch-A-CI

LYHern committed:
Fix style violations

LYHern committed:
Fix JUnit tests

YiHern-Lee committed:
Merge pull request #2 from YiHern-Lee/branch-A-CI

Add support for github CI

LYHern committed:
Add support to undo commands

LYHern committed:
Improve GUI of chat box

LYHern committed:
Improve GUI of chat bot

LYHern committed:
Add screenshot

LYHern committed:
Add user guide

LYHern committed:
Update Javadocs and fix bugs for release

LYHern committed:
Fix Javadoc header comments
